compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp

```diff
@@ -9,11 +9,64 @@
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
 #include "mlir/Dialect/Utils/IndexingUtils.h"
 
+#include "llvm/Support/Debug.h"
+
 namespace mlir::iree_compiler {
 
 using IREE::LinalgExt::EncodingAttr;
 using IREE::LinalgExt::EncodingRole;
 
+std::string str(SmallVector<int64_t> v) {
+  std::stringstream ss;
+  ss << "[ " << v[0];
+  for (int i = 1; i < v.size(); ++i) {
+    ss << " " << v[i];
+  }
+  ss << " ]";
+  return ss.str();
+}
+
+static SmallVector<int64_t> reversed(ArrayRef<int64_t> v) {
+  SmallVector<int64_t> result(v.size());
+  for (unsigned i = 0; i < v.size(); ++i)
+    result[v.size() - 1 - i] = v[i];
+  return result;
+}
+
+static RankedTensorType getTransposedType(RankedTensorType tensorType) {
+  SmallVector<int64_t> transposedShape = reversed(tensorType.getShape());
+  IREE::LinalgExt::EncodingAttr encoding =
+      tensorType.getEncoding()
+          .dyn_cast_or_null<IREE::LinalgExt::EncodingAttr>();
+  if (encoding) {
+    auto transposedRole = encoding.getRole().getValue();
+    if (encoding.getRole().getValue() == IREE::LinalgExt::EncodingRole::LHS) {
+      transposedRole = IREE::LinalgExt::EncodingRole::RHS;
+    }
+    if (encoding.getRole().getValue() == IREE::LinalgExt::EncodingRole::RHS) {
+      transposedRole = IREE::LinalgExt::EncodingRole::LHS;
+    }
+    TypeAttr originalTypeAttr = encoding.getOriginalType();
+    TypeAttr transposedOriginalTypeAttr;
+    if (originalTypeAttr) {
+      RankedTensorType originalType = getTransposedType(
+          originalTypeAttr.getValue().cast<RankedTensorType>());
+      transposedOriginalTypeAttr = TypeAttr::get(originalType);
+    }
+    auto transposedEncoding = IREE::LinalgExt::EncodingAttr::get(
+        encoding.getContext(),
+        IREE::LinalgExt::EncodingRoleAttr::get(encoding.getContext(),
+                                               transposedRole),
+        encoding.getElementTypes(), transposedOriginalTypeAttr,
+        encoding.getMatmulNarrow_N(), encoding.getMatmulNarrow_M(),
+        encoding.getUserIndexingMaps());
```

Comment:
My guess as to what is happening with the unrealized conversion cast is that you are not transposing the indexing maps here. Since the `materializeEncodingFn` uses indexing maps to compute the pack metadata now, you'll need to transpose them here too.

Reply:
oh... i had thought of that but wasn't sure what the UserIndexingMaps were used for and thought they weren't playing a role here. I can try implementing that. I'm curious -- where in my above -debug log are things starting to look like the consequence of not transposing the indexing maps ?

Reply:
You should be able to see something wrong with the `MaterializeEncodingInfo` after `!!!!!! TRANSPOSE !!!!!!`. I'm not sure what metadata it is supposed to have for this case, but you can check that it looks correct

Reply:
This looks correct to me:

```
MaterializeEncodingInfo:
> innerTileSizes[ 4 16 ]
> innerDimsPos[ 0 1 ]
> outerDimsPerm[ 0 1 ]
```

Also in the final `unrealized_conversion_cast` error, the types look correct:

```
/home/benoit/matmul_n4_i8.mlir:2:13: error: failed to legalize unresolved materialization from 'tensor<?x1x16x4xi32>' to 'tensor<1x?x4x16xi32>' that remained live after conversion
  %result = linalg.matmul ins(%lhs, %rhs: tensor<?x?xi8>, tensor<?x4xi8>) outs(%acc: tensor<?x4xi32>) -> tensor<?x4xi32>
            ^
/home/benoit/matmul_n4_i8.mlir:1:1: note: called from
func.func @matmul_n4_i8(%lhs: tensor<?x?xi8>, %rhs: tensor<?x4xi8>, %acc: tensor<?x4xi32>) -> tensor<?x4xi32> {
^
/home/benoit/matmul_n4_i8.mlir:2:13: note: see current operation: %76 = "builtin.unrealized_conversion_cast"(%75) : (tensor<?x1x16x4xi32>) -> tensor<1x?x4x16xi32>
```

Reply:
@Max191 I have pushed a commit implementing the transposition of the indexing maps. Still same result.